infra/Pos/Binary/Infra/Slotting.hs

```diff
@@ -4,24 +4,31 @@ module Pos.Binary.Infra.Slotting
        (
        ) where
 
+import           Universum
+
 import           Data.Time.Units    (Millisecond)
+import           Data.Map           as M
 
-import           Pos.Binary.Class   (Cons (..), Field (..), deriveSimpleBi)
+import           Pos.Binary.Class   (Cons (..), Field (..), Bi (..), deriveSimpleBi)
 import           Pos.Binary.Core    ()
 import           Pos.Core.Timestamp (TimeDiff)
-import           Pos.Core.Types     (EpochIndex)
-import           Pos.Slotting.Types (EpochSlottingData (..), SlottingData (..))
+import           Pos.Slotting.Types (EpochSlottingData (..), SlottingData, getSlottingDataMap, createSlottingDataUnsafe)
 
 deriveSimpleBi ''EpochSlottingData [
     Cons 'EpochSlottingData [
         Field [| esdSlotDuration :: Millisecond |],
         Field [| esdStartDiff    :: TimeDiff    |]
     ]]
 
--- CSL-1122: add a test for serialization of 'SlottingData'
-deriveSimpleBi ''SlottingData [
-    Cons 'SlottingData [
-        Field [| sdPenult      :: EpochSlottingData |],
-        Field [| sdLast        :: EpochSlottingData |],
-        Field [| sdPenultEpoch :: EpochIndex        |]
-    ]]
+instance Bi SlottingData where
+    encode slottingData = encode $ getSlottingDataMap slottingData
+    decode = checkIfSlottindDataValid $ decode
+      where
+        -- We first check if the data we are trying to decode is valid.
+        -- We don't want to throw a runtime error.
+        checkIfSlottindDataValid slottingDataM = do
+            slottingData <- slottingDataM
+            if M.size slottingData < 2
```

Comment:
I see you added more checks to `createSlottingDataUnsafe`, but didn't add them here. It makes sense to create a function which creates `SlottingData` and returns `Either Text SlottingData`.